contracts/modules/cover/Cover.sol

```diff
@@ -173,124 +191,111 @@ contract Cover is ICover, MasterAwareV2 {
-              uint96(coveredAmountInNXM),
-              uint96(premiumInNXM)
-            ));
         }
+        coverChunksForCover[newCoverId].push(
+          CoverChunk(
+            params.coverChunkRequests[i].poolAddress,
+            uint96(coveredAmountInNXM),
+            uint96(premiumInNXM)
+          ));
       }
-      require(amountLeftToCoverInNXM == 0, "Not enough available capacity");
     }
 
-    premiumInAsset = totalPremiumInNXM * tokenPrice / 1e18;
+    premiumInPaymentAsset = totalPremiumInNXM * pool().getTokenPrice(params.paymentAsset) / 1e18;
 
     // make the previous cover expire at current block
     uint32 elapsedPeriod = originalCover.period - remainingPeriod;
-    uint96 updatedOriginalPremium = originalCover.premium * elapsedPeriod / originalCover.period;
-    uint96 carriedPremium = originalCover.premium - updatedOriginalPremium;
+    uint96 updatedOriginalPremium =
+      uint96(originalCover.premium * originalCover.amount / BASIS_PRECISION * elapsedPeriod / originalCover.period);
+    uint96 carriedPremium = originalCover.premium * originalCover.amount - updatedOriginalPremium;
 
     originalCover.period = elapsedPeriod;
-    covers[coverId].premium = updatedOriginalPremium;
+    originalCover.premium = updatedOriginalPremium * uint96(BASIS_PRECISION) / originalCover.amount;
 
-    covers[coverId] = CoverData(
+    uint96 amount = uint96(totalCoverAmountInNXM * tokenPrice / 1e18);
+    covers[params.coverId] = CoverData(
         originalCover.productId,
         originalCover.payoutAsset,
         originalCover.amount + amount,
         uint32(block.timestamp), // start
         remainingPeriod,
-        uint96(premiumInAsset + carriedPremium)
+        uint96((totalPremiumInNXM * tokenPrice / 1e18 + carriedPremium)  * BASIS_PRECISION / (originalCover.amount + amount))
       );
 
     // mint the new cover
     coverNFT.safeMint(msg.sender, newCoverId);
   }
 
-  function increasePeriod(uint coverId, uint32 extraPeriod, uint maxPremiumInAsset) external payable onlyMember {
+  function increasePeriod(uint coverId, uint32 extraPeriod, uint8 paymentAsset, uint maxPremiumInAsset) external payable onlyMember {
 
 
     CoverData memory cover = covers[coverId];
     require(cover.start + cover.period > block.timestamp, "Cover: cover expired");
 
-    uint premiumInAsset = _increasePeriod(coverId, extraPeriod);
+    uint premiumInAsset = _increasePeriod(coverId, extraPeriod, paymentAsset);
     require(premiumInAsset <= maxPremiumInAsset, "Cover: Price exceeds maxPremiumInAsset");
 
     retrievePayment(premiumInAsset, covers[coverId].payoutAsset);
   }
 
-  function _increasePeriod(uint coverId, uint32 extraPeriod) internal returns (uint) {
+  function _increasePeriod(uint coverId, uint32 extraPeriod, uint8 paymentAsset) internal returns (uint) {
 
     CoverData storage cover = covers[coverId];
     CoverChunk[] storage coverChunks = coverChunksForCover[coverId];
```

Comment:
It looks like increasing the amount allows you to add an unlimited amount of pools, while increasing the period only allows you to do so using the same pools. How would this work with unstake requests that diminish the capacity? Wouldn't this lead you to the outcome of not being able to increase the period due to the inability of providing a list of pools other than the existing ones?